frontend/src/metabase/query_builder/components/notebook/lib/steps.js

```diff
@@ -1,8 +1,86 @@
+/* @flow weak */
+
 import StructuredQuery from "metabase-lib/lib/queries/StructuredQuery";
+import type Question from "metabase-lib/lib/Question";
+
+import type { StructuredDatasetQuery } from "metabase/meta/types/Card";
 
 import _ from "underscore";
 
-const STEPS = [
+// This converts an MBQL query into a sequence of notebook "steps", with special logic to determine which steps are
+// allowed to be added at every other step, generating a preview query at each step, how to delete a step,
+// ensuring steps that become invalid after modifying an upstream step are removed, etc.
+
+type StepType =
+  | "data"
+  | "join"
+  | "expression"
+  | "filter"
+  | "summarize"
+  | "sort"
+  | "limit";
+
+type StepDefinition = {
+  // a type name for the step
+  type: StepType,
+  // returns true if the step can be added to the provided query
+  valid: (query: StructuredQuery) => boolean,
+  // returns true if the provided query already has this step
+  active: (query: StructuredQuery) => boolean,
+  // logic to remove invalid clauses that were added by this step from the provided query
+  clean: (query: StructuredQuery) => StructuredQuery,
+  // logic to remove this step from the provided query
+  revert?: (query: StructuredQuery) => StructuredQuery,
+  // if the step can be viewed as multiple steps, how many sub-steps are there
+  // if provided, all the other functions are called with the query and the index of the substep
+  subSteps?: (query: StructuredQuery) => number,
+};
+
+// identifier for this step, e.x. `0:data` (or `0:join:1` for sub-steps)
+type StepId = string;
+
+type Step = {
+  id: StepId,
+  // the step type, corresponds with type in StepDefinition
+  type: StepType,
+  // if there are nested queries, this indicates the level of nested (0 being the root query)
+  stageIndex: number,
+  // if there are sub-steps, this indicates the index of the sub-step
+  itemIndex: number,
+  // is this step currently allowed?
+  valid: boolean,
+  // is this step currently applied?
+  active: boolean,
+  // is this step visible? (if it's active or just added)
+  visible: boolean,
+  // the query for this "stage"
+  query: StructuredQuery,
+  // a query to preview query at this step
+  previewQuery: ?StructuredQuery,
+  // remove this step
+  revert: ?(query: StructuredQuery) => StructuredQuery,
+  // remove invalid clauses set by this step
+  clean: (query: StructuredQuery) => StructuredQuery,
+  // update the query at this step and clean subsequent queries
+  update: (datasetQuery: StructuredDatasetQuery) => StructuredQuery,
+  // any valid actions that can be applied after this step
```

Comment:
These are great comments! They're concise, but they add a lot over the unexplained method names.